### ISSUE TITLE
hedgehog

### ISSUE THREAD

**Issue by tomjaguarpaw** [tomjaguarpaw/tilapia#39]
https://github.com/hedgehogqa/haskell-hedgehog/blob/7030059c089b45429c9ed23f37971e1b5334dc2e/hedgehog/src/Hedgehog/Internal/Property.hs#L355

**Comment by kindaro** [comment]
I suppose this is also a lazy list cargo cult? Maybe a strict vector should be used here.

By the way, [a recent blog post] says 2 things:

* That strict data fields should always be used.

  > Advocate strict data fields by default

  Do you have an argument against this suggestion?

* That a vector should be either strict or unboxed.

  > Oh, and there’s no such thing as a strict, boxed vector. Boo.

  It seems that in the case here we want a strict boxed vector.

Perhaps the Tilapia coalition should voice an opinion on these points.

[a recent blog post]: https://www.snoyman.com/blog/2020/11/haskell-bad-parts-2

**Comment by tomjaguarpaw** [comment]
At some point I will take the time to write down more precisely what I mean by "bang pattern cargo cult". In brief, the bang pattern (or rather strict data field annotation) on `![(PropertyName, Property)]` does almost nothing. It only forces the list to a `(:)` or a `[]`. It doesn't even force the tuple inside the first element (if any) let alone actually force any `PropertyName` or `Property`. Moreover, a `PropertyName` is just a `String` and forcing that would do equally little!

> > Advocate strict data fields by default

> Do you have an argument against this suggestion?

I think this suggestion is good but there's no point having a strict data field if forcing the data type doesn't deeply force it.

**Comment by kindaro** [comment]
Yes, so my proposition is that the right way to go in such cases is to enable the `StrictData` language extension and also to use a strict boxed vector instead of a list. Unfortunately, the default tuple is a lazy constructor even in the presence of `StrictData`, so some sort of strict tuples [_(say this)_][1] should be used.

My message is that the story of the strict data field cargo cult is incomplete until you mention that the usual types that one may use when defining new data constructors, such as tuples and lists, are lazy in all cases and that other, not widely known types should be used.

[1]: https://hackage.haskell.org/package/strict-tuple-0.1.1/docs/Data-Tuple-Strict.html

**Comment by tomjaguarpaw** [comment]
Agreed that it would be good for tilapia to come up with some very clear guidelines for this.

**Comment by tomjaguarpaw** [comment]
Waiting for: more examples of the bang pattern cargo cult.  Then write an article about it.
